Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,684 @@
+carefully read the regular expression or run the regular expression on some
+input to find out. Because regular expressions are oftentimes statically
+available in the source code, there is a missed opportunity to use generics to
+present captures as part of type information to the programmer, and to leverage
+the compiler to infer the type of captures based on a regular expression
+literal. As we propose to introduce declarative string processing capabilities
+to the language and the Standard Library, we would like to explore a type-safe
+approach to regular expression captures.
+
+## Proposed solution
+
+We introduce a generic structure `Regex<Captures>` whose generic parameter
+`Captures` denotes the type of the captured content of such a regular
+expression. With a single generic parameter `Captures`, we make use of tuples to
+represent multiple and nested captures, and are looking to be one of the first
+adopters of variadic generics.
+
+```swift
+let regex = /ab(cd*)(ef)gh/
+// => Regex<(Substring, Substring)>
+if let match = "abcddddefgh".firstMatch(regex) {
+  print(match.captures) // => ("cdddd", "ef")
+}
+```
+
+During type inference for regular expression literals, the compiler infers the
+capture type based on the regular expression's content.  Same for the result
+builder syntax, except that the type inference rules are expressed as method
+declarations in the result builder type.
+
+## Detailed design
+
+### `Regex` type
+
+`Regex` is a structure that represents a regular expression. `Regex` is generic
+over an unconstrained generic parameter `Captures`. Upon a regex match, the
+captured value is available in type `Captures` in the match result.
+
+```swift
+public struct Regex<Captures>: RegexProtocol, ExpressibleByRegexLiteral {
+    ...
+}
+```
+
+### Capture type
+
+In this section, we dive into capture types for regular expression patterns and
+how they compose.
+
+By default, a regular expression literal has type `Regex`. Its generic argument
+`Capture` is its capture type.
+
+#### Basics
+
+Regular expressions without any capturing groups have type `Regex<Void>`, for example:
+
+```swift
+let identifier = /[_a-zA-Z]+[_a-zA-Z0-9]*/  // => `Regex<Void>`
+
+// Result builder DSL equivalent:
+//     let identifier = Regex {
+//         OneOrMore(/[_a-zA-Z]/)
+//         Repeat(/[_a-zA-Z0-9]/)
+//     } // => `Regex<Void>`
+```
+
+#### Capturing group
+
+In regular expression literals, a capturing group is a pattern wrapped by a pair
+of parentheses. A capturing group's capture type is reflected in the result
+type's `Capture` generic argument. A capturing group's corresponding capture type
+is `Substring`.
+
+```
+pattern ::= '(' pattern ')'
+```
+
+```swift
+let graphemeBreakLowerBound = /([0-9a-fA-F]+)/ // => `Regex<Substring>`
+
+// Result builder DSL equivalent
+//     let graphemeBreakLowerBound = OneOrMore(CharacterClass.hexDigit).capture()
+//     // `.Captures == Substring`
+```
+
```

Comment:
How do you feel about this alternate formulation? (rendered version of above commit suggestion)

#### Capturing group: `(...)`

In regular expression literals, a capturing group saves the portion of the input matched by its contained regex. A capturing group's `Capture` type is `Substring`.

```swift
let graphemeBreakLowerBound = /([0-9a-fA-F]+)/ // => `Regex<Substring>`

// Result builder DSL equivalent
//     let graphemeBreakLowerBound = OneOrMore(CharacterClass.hexDigit).capture()
```

Reply:
Good suggestion!  I'll update it.

Reply:
Updated as suggested.  I initially thought about using `(...)` earlier but I didn't because it could be confused with the `.` metacharacter. Formal grammar probably is the clearest way to describe it. Nonetheless, I agree `(...)` helps reduce the emphasis on syntax.